core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,254 @@
+
+from ..common.seed import remove_ed25519_prefix
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Dict, List, Optional, Tuple, Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = Dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = Dict[int, bytes]
+    CatalystRegistration = Dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+CATALYST_REGISTRATION_KEY = 61284
+CATALYST_REGISTRATION_SIGNATURE_KEY = 61285
+
+
+def validate_auxiliary_data(
+    auxiliary_data: Optional[CardanoTxAuxiliaryDataType],
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        if not auxiliary_data.blob:
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # currently we only support CATALYST_REGISTRATION
+        if (
+            auxiliary_data.blob
+            or not auxiliary_data.metadata
+            or auxiliary_data.metadata.type != CardanoMetadataType.CATALYST_REGISTRATION
+        ):
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_catalyst_registration_parameters(
+            auxiliary_data.metadata.catalyst_registration_parameters
+        )
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_auxiliary_data_blob(auxiliary_data_blob: bytes) -> None:
+    try:
+        # this also raises an error if there's some data remaining
+        cbor.decode(auxiliary_data_blob)
+    except Exception:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_catalyst_registration_parameters(
+    catalyst_registration_parameters: Optional[
+        CardanoCatalystRegistrationParametersType
+    ],
+) -> None:
+    if not catalyst_registration_parameters:
+        raise INVALID_AUXILIARY_DATA
+
+    if len(catalyst_registration_parameters.voting_public_key) != PUBLIC_KEY_LENGTH:
+        raise INVALID_AUXILIARY_DATA
+
+    if not SCHEMA_STAKING_ANY_ACCOUNT.match(
+        catalyst_registration_parameters.staking_path
+    ):
+        raise INVALID_AUXILIARY_DATA
+
+    address_parameters = catalyst_registration_parameters.reward_address_parameters
+    if address_parameters.address_type == CardanoAddressType.BYRON:
```

Comment:
worth clarifying in docs that byron addresses are not supported for catalyst registration

Reply:
Done in https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/08c938e56c6a49e9d108095339eec63141bf6799.